### PR TITLE
Docs to bring the External Filter definition up to parity with AuthService [apro #310]

### PR DESCRIPTION
**Pull request by LukeShu**
Go review/merge https://github.com/datawire/ambassador/pull/2221 first.  Then fiddle with the base branch of this PR to flush GitHub's cache, so the diff is useful.

## Description

These are the doc changes going along with https://github.com/datawire/apro/pull/892.

I thought that this would be a small amount of work, because I could just copy/paste things from auth-service.md.  But it turns out that improving auth-service.md to a point where it is copy/pastable was a significant chunk of work.

## Related Issues

Issue: https://github.com/datawire/apro/issues/310
PR: https://github.com/datawire/apro/pull/892

## Other

They are made here, instead of in ambassador-docs.git, because it shouldn't be in the published docs until we do a release incorporating https://github.com/datawire/apro/pull/892.

### CODE REVIEW

File: docs/reference/services/auth-service.md [hunk 1]
```diff
@@ -49,101 +76,25 @@ spec:
...
-- The external auth service can issue a 301 `Redirect` to divert the client into an OAuth or OIDC authentication sequence.
+## Canarying Multiple AuthServices
 
-Finally, if Ambassador Edge Stack cannot reach the auth service at all, it will return a HTTP 503 status code to the client.
+You may create multiple `AuthService` manifests to round-robin authentication requests among multiple services. **Note well that all services must use the same `path_prefix` and header definitions;** if you try to have different values, you'll see an error in the diagnostics service, telling you which value is being used.
 
 ## Configuring Public Mappings
 
-Authentication can be disabled for a mapping by setting `bypass_auth` to `true`. This will tell Ambassador Edge Stack to allow all requests for that mapping through without interacting with the external auth service.
+An `AuthService` can be disabled for a mapping by settingn `bypass_auth` to `true`. This will tell Ambassador to allow all requests for that mapping through without interacting with the external auth service.
```
Comment (scoyle391) [inline docs/reference/services/auth-service.md]: settingn > setting

File: docs/reference/services/ext_authz.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+## The `ext_authz` Protocol
+
+By design, the `ext_authz` protocol used by [the `AuthService`](../auth-service) and by [`External` `Filters`](../../filter-reference) is highly flexible. The authentication service is the first external service invoked on an incoming request (e.g., it runs before the rate limit filter). Because the logic of authentication is encapsulated in an external service, you can use this to support a wide variety of use cases. For example:
+
+* Supporting traditional SSO authentication protocols, e.g., OAuth, OpenID Connect, etc.
```
Comment (scoyle391) [inline docs/reference/services/ext_authz.md]: Try to have the same verb form for these bullets. So either all ending in -ing or in the infinitive, etc, for consistency

File: docs/reference/services/ext_authz.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+* Support HTTP basic authentication (sample implementation available [here](https://github.com/datawire/ambassador-auth-httpbasic)).
+* Only authenticating requests that are under a rate limit, and rejecting authentication requests above the rate limit.
+* Authenticating specific services (URLs), and not others.
+
+For each request, the external auth service may either
+ 1. return a direct HTTP *response*, intended to be sent back to the requesting HTTP client (normally *denying* the request from being forwarded to the upstream backend service); or
+ 2. return a modification to make to the HTTP *request* before sending it to the upstream backend service (normally *allowing* the request to be forwarded to the upstream backend service with modifications).
+
+The external auth service receives information about every request through Ambassador and must indicate whether the request is to be allowed, or not.  If not, the external auth service provides the HTTP response which is to be handed back to the client.  A potential control flow for Authentication is shown below.
```
Comment (scoyle391) [inline docs/reference/services/ext_authz.md]: as a general rule, if you're referencing something, don't say above or below. please include a name of the section; if the section does not have a name, give it one. :D

Comment (scoyle391) [reply docs/reference/services/ext_authz.md]: or if this is referring to the image, ignore the bit (but take the advice?)

File: docs/reference/services/ext_authz.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+```
+PUT /path/to/service HTTP/1.1
+Host: myservice.example.com:8080
+User-Agent: curl/7.54.0
+Accept: */*
+Content-Type: application/json
+Content-Length: 27
+
+{ "greeting": "hello world!", "spiders": "OMG no" }
```
Comment (scoyle391) [inline docs/reference/services/ext_authz.md]: doge spider

File: docs/doc-links.yml [hunk 5]
```diff
@@ -209,6 +209,8 @@
         link: /reference/services/access-control
       - title: AuthService Plugin
         link: /reference/services/auth-service
+      - title: Log Plugin
```
[the remark this answers is not in the document]
Comment (LukeShu) [reply docs/doc-links.yml]: I... didn't change this, that was just part of #2221

File: docs/reference/core/crds.md [hunk 6]
```diff
@@ -33,11 +33,12 @@ spec:
   service: luatest-http
 ```
 
-Note that the `namespace` must be declared in the `metadata`, but if needed, `ambassador_id` must be declared in the `spec`.
+Note that the `namespace` must be declared in the `metadata`, but if needed, `ambassador_id` must be declared in the `spec`.  
+You can run `kubectl get ambassador-crds` to see the CRDs you've created to configure the Ambassador Edge Stack.
```
[the remark this answers is not in the document]
Comment (LukeShu) [reply docs/reference/core/crds.md]: I... didn't change this, that was just part of #2221

Also, your suggested "should be" is factually incorrect.

File: docs/reference/services/ext_authz.md (hunk 2 above)
Comment (LukeShu) [reply docs/reference/services/ext_authz.md]: Making me reword things that I just cut/pasted from one file to another. :)

File: docs/reference/services/ext_authz.md (hunk 3 above)
Comment (LukeShu) [reply docs/reference/services/ext_authz.md]: It's referring to the image.

File: docs/reference/services/ext_authz.md (hunk 2 above)
Comment (scoyle391) [reply docs/reference/services/ext_authz.md]: sigh

File: docs/reference/core/crds.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (scoyle391) [reply docs/reference/core/crds.md]: k!

File: docs/doc-links.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (scoyle391) [reply docs/doc-links.yml]: okay, commented there instead

File: docs/reference/services/ext_authz.md (hunk 2 above)
Comment (LukeShu) [reply docs/reference/services/ext_authz.md]: Fixed, they all use the -ing form now.

**Comment by LukeShu** [comment]
I've made some of @scoyle391's requested changes here, and some in a separate https://github.com/datawire/ambassador-docs/pull/268.

File: docs/reference/filter-reference.md [hunk 7]
```diff
@@ -35,9 +35,7 @@ Currently, the Ambassador Edge Stack supports four filter types: `External`, `JW
 
 ### Filter Type: `External`
 
-The `External` filter type exposes the Ambassador Edge Stack `AuthService` interface to external authentication services. This is useful in a number of situations, e.g., if you have already written a custom `AuthService`, but also want to use other filters.
-
-The `External` filter looks very similar to an `AuthService` annotation:
+The `External` filter calls out to an external service speaking the [`ext_authz` protocol][], providing a highly flexible interface to plug in your own authentication, authorization, and filtering logic.
```
Comment (kflynn) [inline docs/reference/filter-reference.md]: What does `[ext_authz protocol][]` mean in Markdown? :joy:

Comment (scoyle391) [reply docs/reference/filter-reference.md]: @LukeShu did you mean to reference something here?

**Comment by kflynn** [comment]
Oh, she already approved. Merging!

**Pull request merged**